Added a time saver option to allow players to instantly put items away

1SDANi opened the pull request:
Expanded the range of damage multiplication to 80 (20 hearts of damage for each quarter heart) so players can play a sudden death challenge

1SDANi committed:
Added a time saver option to allow players to instantly put items away

Expanded the range of damage multiplication to 80 (20 hearts of damage for each quarter heart) so players can play a sudden death challenge

briaguya0 commented:
would you mind screenshotting the slider for damage multiplier with the new change? gut feeling is it'll be hard to set the slider to 2x or 4x now that it goes all the way up to 80x

vaguerant commented:
That seems like a bigger problem with sliders in general. Maybe the slider implementation should be improved across the board? An obvious way to do it would be to have something like this:

`[ < ] . . . . . . . . . . |_____| . . . . . [ > ]`

Pardon the miserable ASCII art, but clickable buttons that go one unit left/right at each end of a slider seem like they'd improve usability without cluttering the menu much.

1SDANi commented:
Control clicking lets you set slider values via the keyboard. It's global behaviour on sliders.
I will add a tooltip specifying that you can control click though.

vaguerant commented:
> Control clicking lets you set slider values via the keyboard. It's global behaviour on sliders. I will add a tooltip specifying that you can control click though.

The thing is that Ctrl-clicking renders sliders moot. It may as well not be a slider input at all if you have to Ctrl-click it and type in a value manually to get sensible values out of it--that's just a textbox with extra steps. To be clear, I'm not at all critiquing your use of a slider, just the way sliders work right now in SoH.

The strength of a slider is making large, approximate changes. Think scrolling a web page: you don't need precise output from the scrollbar, pretty close is good enough. And notably, you have the freedom to adjust the final result of a scroll with e.g. cursor keys or the buttons that sit at the top/bottom of the scroll bar, to dial in on exactly what you're trying to see.

The way we're using sliders in SoH--to select a specific value from a range, with no way to make finer adjustments--is diametrically opposed to what sliders are good at. If you want 30 FPS on the interpolation slider but you keep skipping between 29 and 31, the slider is just wasting your time. At that point, Ctrl-clicking is just an admission that the slider is useless.

1SDANi commented:
If I could make the slider a text field, I would.

vaguerant commented:
Maybe `ImGui::InputScalar` is worth trying out? That's used in the Save Editor to get these things:

![InputScalar demo](https://user-images.githubusercontent.com/5259025/177607241-ed4728c2-ccfa-4757-848f-46657360e05e.png)

https://github.com/HarbourMasters/Shipwright/blob/537722d57a1fdd2fed24ccbab84179eb97271162/soh/soh/Enhancements/debugger/debugSaveEditor.cpp#L417-L418

1SDANi commented:
ooh, that looks perfect!

briaguya0 commented:
i guess my bigger question is: will people want to use any of the values between like 4 and 80? seems to me the idea is to have a "one hit ko" mode, so maybe just a slider with reasonable values and then a checkbox next to it that overrides the slider and sets the value to 80?

1SDANi commented:
probably not, but I think I'd prefer to have 1 scalar input cvar than 1 slider and 1 checkbox

briaguya0 commented:
i'm just basing this idea on what n64 randomizer does
![image](https://user-images.githubusercontent.com/70942617/177611727-e72087fc-f2da-4fb2-b364-2f3942d29ffc.png)

being able to pick a 79x damage multiplier when it will be functionally identical to setting an 80x multiplier in 99% of scenarios feels like bad UX to me regardless of the input widget used

1SDANi commented:
How is giving less options better? Yes, most people won't use most options over 4x, but the user experience isn't hurt by allowing players to input whatever positive number they want into the scalar. I see no reason to artificially constrain what players can input into the scalar.

briaguya0 commented:
the UX *is* hurt if you aren't using a keyboard

changing all the way up to 80x will take 80 clicks on the scalar

1SDANi commented:
I mean, that's more a problem with scalars not allowing you to input text with a gamepad

briaguya0 commented:
if 99% of players will pick one of 5 values, giving 100% of players 80 choices (with a less intuitive UI than the 5 could have) seems like a bad move

at the end of the day it's not my call to make, just wanted to make sure to voice my opinion on the matter

1SDANi commented:
I fail to see how an input field is less intuitive than a drop down menu. Even on console, input fields are consistently used when players input numerical values such as their phone number or credit card info, as opposed to say, 3 drop downs each with 10 options for the 3 digits on the back of your credit card.

briaguya0 commented:
with credit cards and phone numbers there are trillions of possible values people would use, so it makes sense to have it be a numerical text input of some sort

with this damage multiplier, nobody is going to use the vast majority of those values

unrelated: if you want it to still be OHKO when the player has double defense you'll need to up the max to 160

1SDANi commented:
the only difference between a text field on console and a drop down is that players have to use the X axis in addition the Y axis and confirm button.
good catch

MelonSpeedruns commented:
I am not approving this PR unless it is changed to a combo box, similar to the one in the N64 Rando.
I think having only the options in the screenshot @briaguya-ai sent above would be perfect.

1SDANi commented:
fair enough, though it's still outright worse in terms of UX

MelonSpeedruns commented:
You already have been proven that it is not the case. There's a reason Starbucks doesn't ask you how many millilitres of coffee you want, most people wouldn't know what to answer. Same thing here, and it would be especially punishing for people that either don't know the game well enough, or people that want the options to be quickly dealt with.

I've heard so many times before that Dolphin was way too complicated for a casual user to use, with all these options, and they were lost. If you're asking the user to do maths just in order for them to set their damage value just right, there's an issue.

It's also why games don't have a "Custom" difficulty option, where the user can specifically set the damage values & whatever else the game has. Instead, most games have a "Easy", "Normal", and "Hard" difficulty modes, so why try to reinvent the wheel?

1SDANi commented:
I already have proven that it is the case, any assertion to the contrary is ignoring solid evidence.
Millimeters and cups are two separate measuring amounts.

having given it more thought, the above options are clearly not enough. There needs to be at the very least a 20, 40, or 80x option for a ohko option wherein double damage is not useless in addition to a 160x option wherein it is, and ideally the other options as well.

Again, this is the exact sort of thing where it's best left up to the player. If someone walks into a Starbucks and asks for 5 lattes, no server is going to say they only serve orders containing 1, 2, or 4 lattes.

briaguya0 commented:
new idea, go back to the slider, but instead of having it determine the multiplier directly it determines the exponent

so we'd have gDamageMul set to
2^0 = 1x
2^1 = 2x
2^2 = 4x
2^3 = 8x
2^4 + 16x
2^5 = 32x
2^6 = 64x
2^7 = 128x
2^8 = 256x (OHKO)

this gives us 9 spots on a slider with meaningful differences in player experience

1SDANi commented:
that seems like a fair compromise

1SDANi committed:
The damage sliders are now comboboxes
Damage multiplication is now exponential instead of incremental

The FPS slider is now a text input

1SDANi commented:
Sorry it took so long to make these changes, I've been rather busy spending time with family the last few days

1SDANi committed:
updated to latest develop

1SDANi commented:
oh, forgot to mention, but this also makes the fps slider a text input field as previously discussed

PurpleHato commented:
I think the FPS as a slider with incremental of 20 is still a better solution and the CTRL+Click on it to input your desired value of necessary

But that's my two cents

1SDANi commented:
Ooh, that's an even better idea, I'mma implement it now then.

PurpleHato commented:
> Ooh, that's an even better idea, I'mma implement it now then.

Awesome thanks!

1SDANi commented:
Um, bad news, I'm not sure sliders allow for incrementing in steps atm, and I don't think I understand the code well enough to implement it.

PurpleHato commented:
> Um, bad news, I'm not sure sliders allow for incrementing in steps atm, and I don't think I understand the code well enough to implement it.

I think we can create a new slider type based on the non step one (like we did with enhancedXXX)
This looks like a good starting point https://discourse.dearimgui.org/t/is-there-a-easy-way-to-step-sliders/239/5

jbodner09 commented:
Another option for the FPS would just be to do the math wherever the FPS variable is used, and have the slider still be continuous integer values of 1 through 12.  Then, have the FPS variable be (cvar_val * 20), so the minimum value would be 20 and the maximum would be 240 (instead of the current 250, which seems arbitrary to me anyway).

PurpleHato commented:
> Another option for the FPS would just be to do the math wherever the FPS variable is used, and have the slider still be continuous integer values of 1 through 12. Then, have the FPS variable be (cvar_val * 20), so the minimum value would be 20 and the maximum would be 240 (instead of the current 250, which seems arbitrary to me anyway).

indeed, that's also a very valid solution ^-^

Kenix3 commented:
Can we split PRs in the future? There's no reason that the damage multiplier fixes need to be tied to a quick item put away enhancement.

Kenix3 reviewed (changes requested):
Please split the PR. Move the damage multipliers into their own PR. Also, quadruple is mispelled. Thankyou!

1SDANi commented:
sorry, was intending to bump out a lot of different issues and enhancements in one big PR but it ended up just being two features. Will split and going forwards will keep PRs to a single change

1SDANi committed:
Splitting into solely being a quick putaway PR

1SDANi committed:
updated to latest develop

1SDANi committed:
Removed some leftover code from the split

dcvz commented on libultraship/libultraship/ImGuiImpl.cpp:
```diff
@@ -1169,6 +1170,15 @@ namespace SohImGui {
 
                     if (ImGui::SliderInt("##FPSInterpolation", &val, 20, 250, "", ImGuiSliderFlags_AlwaysClamp))
                     {
+                        if (val > 250)
```
I think you can use clamp for this:

```c++
val = math::clamp(val, 20.0f, 250.0f);
```

dcvz reviewed (approved):
LGTM 👍
